### PR TITLE
stream info: add listener info

### PR DESCRIPTION
**Pull request by kyessenov**
Commit Message: Expose listener info in stream info.
Additional Description: Add relevant accessors in  the access log and CEL attributes.
Risk Level: low
Testing: added
Docs Changes: yes
Release Notes: yes
Fixes: https://github.com/envoyproxy/envoy/issues/24303

CC @zirain @wbpcode

### CODE REVIEW

File: envoy/network/socket.h [hunk 1]
```diff
@@ -324,6 +331,11 @@ class ConnectionInfoSetter : public ConnectionInfoProvider {
    * @param filter_chain_info the filter chain info provider backing this socket.
    */
   virtual void setFilterChainInfo(FilterChainInfoConstSharedPtr filter_chain_info) PURE;
+
+  /**
+   * @param listener_info the listener info provider backing this socket.
+   */
+  virtual void setListenerInfo(std::shared_ptr<const ListenerInfo> listener_info) PURE;
```
Comment (wbpcode) [inline envoy/network/socket.h:338]: nit: `ListenerInfoConstSharedPtr`.

Comment (kyessenov) [reply envoy/network/socket.h:338]: This is because I can't import `listener.h` due to a header cycle.

File: envoy/network/listener.h [hunk 2]
```diff
@@ -238,9 +240,9 @@ class ListenerConfig {
   virtual const std::string& name() const PURE;
 
   /**
-   * @return ListenerInfo& description of the listener.
+   * @return ListenerInfoConstSharedPtr& description of the listener.
    */
-  virtual const ListenerInfo& listenerInfo() const PURE;
+  virtual const ListenerInfoConstSharedPtr& listenerInfo() const PURE;
```
Comment (wbpcode) [inline envoy/network/listener.h:245]: Is this change necessary?

File: source/extensions/filters/common/expr/context.h [hunk 3]
```diff
@@ -89,6 +89,8 @@ constexpr absl::string_view RouteName = "route_name";
 constexpr absl::string_view RouteMetadata = "route_metadata";
 constexpr absl::string_view UpstreamHostMetadata = "upstream_host_metadata";
 constexpr absl::string_view FilterChainName = "filter_chain_name";
+constexpr absl::string_view ListenerMetadata = "listener_metadata";
```
Comment (zirain) [inline source/extensions/filters/common/expr/context.h:92]: In what consideration not to use `listener.metadata` and `listener.direction`?

**Review by wbpcode — COMMENTED** [review]
This enhancement look good to me. Two comments are added first to start review.

File: envoy/network/listener.h (hunk 2 above)
Comment (kyessenov) [reply envoy/network/listener.h:245]: Yeah, I want to use shared ptr since it's difficult to ensure references are safe and listener info is already stored as a shared ptr.

File: source/extensions/filters/common/expr/context.h (hunk 3 above)
Comment (kyessenov) [reply source/extensions/filters/common/expr/context.h:92]: You mean Wasm attributes? We should recommend the standard `xds` attributes once we finish replacing them, and then deprecate those one-off cases.

File: envoy/network/listener.h (hunk 2 above)
Comment (wbpcode) [reply envoy/network/listener.h:245]: Sorry, I actually didn't get it. The listener info is owned by the factory context and will always have longer lifetime than the `ListenerConfig`. And I will shared the ownership with the connection info provider which ensure I can always access it by the valid connection info provider. 

So, I didn't see the necessity to change this method, except you want to get the ownership of listener info by this method (but I didn't see related usage in this PR).

File: source/extensions/filters/common/expr/context.h (hunk 3 above)
Comment (zirain) [reply source/extensions/filters/common/expr/context.h:92]: My bad, I misunderstood.

File: envoy/network/listener.h (hunk 2 above)
Comment (kyessenov) [reply envoy/network/listener.h:245]: I'm not confident `ActiveTcpListener` stream doesn't outlive `ListenerConfig`. The relationship stored is via direct pointers, and it's guaranteed to exist at the connection acceptance, but not for the whole duration of a stream, no?

**Review by abeyad — APPROVED** [review]
/lgtm api

File: envoy/network/listener.h (hunk 2 above)
Comment (wbpcode) [reply envoy/network/listener.h:245]: AFAIK, the drain manager will ensure all connection be closed before actually destroy a filter chain in the main thread. I am not very sure because long time has gone after last time I read related code. 

But it doesn't matter. Note the connection info will own the listener info also now and the connection info will be owned by the `StreamInfo` which have same lifetime with the related stream.

File: source/common/network/socket_impl.h [hunk 4]
```diff
@@ -93,6 +100,7 @@ class ConnectionInfoSetterImpl : public ConnectionInfoSetter {
   std::string ja3_hash_;
   absl::optional<std::chrono::milliseconds> round_trip_time_;
   FilterChainInfoConstSharedPtr filter_chain_info_;
+  ListenerInfoConstSharedPtr listener_info_;
```
Comment (wbpcode) [inline source/common/network/socket_impl.h:103]: @kyessenov See, the connection info own the listener info and connection info will owned by the `StreamInfo`. So, I didn't get you concern. Do I missed something?

Comment (kyessenov) [reply source/common/network/socket_impl.h:103]: Can `StreamInfo` outlive `ListenerImpl`? Let me try to analyze the ownership, but it's definitely not obvious without knowing the flow control.

Comment (kyessenov) [reply source/common/network/socket_impl.h:103]: OK, yeah, the "shared" Listener parts should be carried over by the completion until all workers stop handling streams [here](https://github.com/envoyproxy/envoy/blob/main/source/common/listener_manager/listener_manager_impl.cc#L852). So we don't have to use a shared pointer.

File: envoy/network/listener.h (hunk 2 above)
Comment (kyessenov) [reply envoy/network/listener.h]: Thanks, you are right, we can simplify this PR a lot if we assume lifetime works.

File: source/common/network/socket_impl.h [hunk 5]
```diff
@@ -93,6 +98,7 @@ class ConnectionInfoSetterImpl : public ConnectionInfoSetter {
   std::string ja3_hash_;
   absl::optional<std::chrono::milliseconds> round_trip_time_;
   FilterChainInfoConstSharedPtr filter_chain_info_;
+  OptRef<const ListenerInfo> listener_info_;
```
Comment (wbpcode) [inline source/common/network/socket_impl.h:101]: hmmm, the premise of my previous comments is that connection info provider will own a copy of listener info shared pointer.

So, I think the listenerInfo() of ListenerConfig need to change it's signature. But seems if we don't change the signature, there is no way to store a listener info shared pointer to the connection info provider.

Seems I am wrong, my bad, again.

Comment (wbpcode) [reply source/common/network/socket_impl.h:101]: Maybe revert to your previous implementation would be better (use a shared pointer here).

Sorry again. 😞

Comment (kyessenov) [reply source/common/network/socket_impl.h]: No problem, reverted. It was worth exploring the other option.

File: docs/root/intro/arch_overview/advanced/attributes.rst [hunk 6]
```diff
@@ -197,14 +199,14 @@ In addition to all above, the following extra attributes are available to Wasm e
...
    plugin_vm_id, string, Plugin VM ID
-   node, :ref:`Node<envoy_v3_api_msg_config.core.v3.node>`, Local node description
-   cluster_name, string, Upstream cluster name
-   cluster_metadata, :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`, Upstream cluster metadata
-   listener_direction, int, Enumeration value of the :ref:`listener traffic direction<envoy_v3_api_field_config.listener.v3.Listener.traffic_direction>`
-   listener_metadata, :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`, Listener metadata
-   route_name, string, Route name
-   route_metadata, :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`, Route metadata
-   upstream_host_metadata, :ref:`Metadata<envoy_v3_api_msg_config.core.v3.metadata>`, Upstream host metadata
+   node, :ref:`Node<envoy_v3_api_msg_config.core.v3.node>`, Local node description. DEPRECATED: please use `xds` attributes.
```
Comment (wbpcode) [inline docs/root/intro/arch_overview/advanced/attributes.rst:202]: Seems `node` is not supported in the xds attributes from above docs?

File: source/server/factory_context_impl.cc [hunk 7]
```diff
@@ -6,13 +6,9 @@ namespace Server {
 FactoryContextImplBase::FactoryContextImplBase(
     Server::Instance& server, ProtobufMessage::ValidationVisitor& validation_visitor,
     Stats::ScopeSharedPtr scope, Stats::ScopeSharedPtr listener_scope,
-    ListenerInfoConstSharedPtr listener_info)
+    const Network::ListenerInfoConstSharedPtr& listener_info)
     : server_(server), validation_visitor_(validation_visitor), scope_(std::move(scope)),
-      listener_scope_(std::move(listener_scope)), listener_info_(std::move(listener_info)) {
-  if (listener_info_ == nullptr) {
-    listener_info_ = std::make_shared<ListenerInfoImpl>();
-  }
```
Comment (wbpcode) [inline source/server/factory_context_impl.cc:14]: nit: `ASSERT(listener_info_ == nullptr);` to avoid unexpected regression.

**Review by wbpcode — COMMENTED** [review]
LGTM overall. It is almost done. Only two minor comments about the docs and an optional ASSERT.

File: source/server/factory_context_impl.cc (hunk 7 above)
Comment (kyessenov) [reply source/server/factory_context_impl.cc:14]: Done.

File: docs/root/intro/arch_overview/advanced/attributes.rst (hunk 6 above)
Comment (kyessenov) [reply docs/root/intro/arch_overview/advanced/attributes.rst:202]: Good catch, the PR just went in, and fixed the missing doc.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.
